src/autopas/utils/PatternBenchmark.h

```diff
@@ -0,0 +1,340 @@
+    using ParticleType = Particle_T;
+    using CellType = autopas::FullParticleCell<ParticleType>;
+
+    const double cutoff{functor.getCutoff()};  // is also the cell size
+
+    std::vector<long> times{};
+    times.reserve(repetitions);
+    functor.setVecPattern(vecPattern);
+
+    for (int n = 0; n < repetitions; ++n) {
+      // define scenario
+      const std::pair<size_t, size_t> numParticlesPerCell{firstnumParticles, secondnumParticles};
+
+      // repeat the whole experiment multiple times and average results
+      std::pair<CellType, CellType> cells{};
+
+      initialization<Functor_T, CellType, ParticleType>(functor, cells, numParticlesPerCell, cutoff, hitRate);
+      for (size_t iteration = 0; iteration < iterations; ++iteration) {
+        // actual benchmark
+        applyFunctorPair(functor, cells, timer, newton3);
+      }
+      // print particles to CSV for checking and prevent compiler from optimizing everything away.
+      csvOutput(functor, cells);
+      times.push_back(timer["Functor"].getTotalTime());
+      for (auto &[_, t] : timer) {
+        t.reset();
+      }
+    }
+    unsigned long sum = std::accumulate(times.begin(), times.end(), static_cast<long>(0));
+    return sum;
+  }
+  /**
+   * helper method to turn vectorization pattern enum into string
+   *
+   * @param vecPattern vectorization pattern
+   * @return string
+   */
+  inline std::string checkVecPattern(const autopas::VectorizationPatternOption::Value vecPattern) {
+    if (vecPattern == autopas::VectorizationPatternOption::Value::p1xVec) {
+      return "p1xVec";
+    } else if (vecPattern == autopas::VectorizationPatternOption::Value::p2xVecDiv2) {
+      return "p2xVecDiv2";
+    } else if (vecPattern == autopas::VectorizationPatternOption::Value::pVecDiv2x2) {
+      return "pVecDiv2x2";
+    } else {
+      return "pVecx1";
+    }
+  }
+  /**
+   * This method calculates the benchmarkSize x benchmarkSize pattern benchmark to determine the optimal vector pattern
+   * for some given combination of SoA sizes.
+   * @tparam Functor_T Functor type for benchmark.
+   * @tparam Particle_T Particle type in benchmark.
+   * @param functor functor that is used in benchmark.
+   * @param newton3 true if newton3 optimization is applied
+   * @return array with optimal vector pattern for different combinations of SoA sizes.
+   */
+  template <class Functor_T, class Particle_T>
+  std::array<autopas::VectorizationPatternOption::Value, _benchmarkSize * _benchmarkSize> calculateVecPatternMap(
+      Functor_T &functor, bool newton3) {
+    using PatternType = autopas::VectorizationPatternOption::Value;
+    std::vector<PatternType> patterns = {PatternType::p1xVec, PatternType::p2xVecDiv2, PatternType::pVecDiv2x2,
+                                         PatternType::pVecx1};
+    std::array<std::array<unsigned long, 4>, _benchmarkSize * _benchmarkSize> allResults{};
```

Comment:
```suggestion
    using PatternType = VectorizationPatternOption;
    const auto patterns = []() {
      const auto patternsSet = VectorizationPatternOption::getAllOptions();
      return std::vector(patternsSet.begin(), patternsSet.end());
    }();

    constexpr size_t numPatterns = 4;
    if (patterns.size() != numPatterns) {
      utils::ExceptionHandler::exception("PatternBenchmark assumes there are only {} vectorization patterns, but"
      "there is actually {}!", numPatterns, patterns.size());
    }

    std::array<std::array<unsigned long, numPatterns>, _benchmarkSize * _benchmarkSize> allResults{};
```
What this does is

- A) Use `VectorizationPatternOption` instead of `VectorizationPatternOption::Value` as described elsewhere
- B) Generates the `std::vector` from `getAllOptions`.
- C) Checks that there are, in fact, 4 patterns. If there are not, this should throw an error alerting the developer to the fact that this should be updated.

It's a bit ugly, and of course, the cleanest solution would be to refactor `Option` so that we can get more compile-time information, but at least it will loudly cause problems if somebody breaks your code later.

Reply:
As Luis also uses the VectorizationPatternOption::Value type for his pattern selection logic, I have decided to use the VectorizationPatternOption type only for the getAllOptions step and for the to_string() method and just do a simple convert back to or from the enum type when necessary. This seems cleaner in my eyes, as I only require the enum value and not any option functionality elsewhere.